### PR TITLE
Fix 2087, 2088, and 2089 by properly setting parameters in rotation-commands and not using command feedback when relying on scratch-vm for motor timing

### LINKED ISSUE #2087: BOOST extension: can't change motor direction while it's running
### Expected Behavior

While a motor is running, having been started in any of the ways available (motor on, on for seconds, or on for rotations), it should be possible while it is running to change the direction of the motor using the "set motor direction" block. In the case of "on for seconds", the motor should continue running for the remainder of the time. In the case of "on for rotations", the motor should (at least approximately) complete the rotations.
 
### Actual Behavior

Not currently working.

### LINKED ISSUE #2088: BOOST extension: can't change motor speed while it's running
### Expected Behavior

While a motor is running, having been started in any of the ways available (motor on, on for seconds, or on for rotations), it should be possible while it is running to change the speed of the motor using the "set motor speed" block. In the case of "on for seconds", the motor should continue running for the remainder of the time. In the case of "on for rotations", the motor should (at least approximately) complete the rotations.

### Actual Behavior

Not currently working.

### LINKED ISSUE #2089: BOOST extension: default motor speed should be 50%?
In the EV3 extension, we set the initial value for the power of the motors to 50%, instead of 100% (i.e. not the default value for the input of the block, but the actual value set internally when the extension loads). We should consider doing the same for the BOOST. 

This would make the initial experience of the motor moving a bit more gentle. It also gives you room for an increase, so that setting it to 100% actually causes it to change.

### PR DESCRIPTION
**Pull request by knandersen**
Resolves #2087 - Can't change motor direction while it's running
Resolves #2088 - Can't change motor power while it's running
Resolves #2089 - Default motor speed should be 50%?

Clarifies the renaming aspect mentioned in #2091

### CODE REVIEW

**Comment by evhan55** [comment]
This tests ok for me after the latest commit!

- Can change direction on motor for all three motor on blocks (`seconds/rotations/on`)
- Can change power on motor for all three motor blocks (`seconds/rotations/on`)
- Motor default is at 50%

File: src/extensions/scratch3_boost/index.js [hunk 1]
```diff
@@ -590,15 +594,16 @@ class Boost {
...
+            previousColor: BoostColor.NONE
         };
 
         /**
          * An array of values from the Boost Vision Sensor.
          * @type {Array}
          * @private
          */
-        this._colorBucket = [];
+        this._colorSamples = [];
```
Comment (evhan55) [inline src/extensions/scratch3_boost/index.js:606]: Does this change belong in another PR, or should the issue this resolves be mentioned in the comment for this PR?

**Comment by evhan55** [comment]
There is some code in this PR related to color sensing, should it be mentioned in the PR comment and reference an issue so I can test it?

File: src/extensions/scratch3_boost/index.js (hunk 1 above)
Comment (knandersen) [reply src/extensions/scratch3_boost/index.js:606]: Yeah thanks for pointing that out @evhan55. It's from #2091!

File: src/extensions/scratch3_boost/index.js [hunk 2]
```diff
@@ -1926,10 +1934,10 @@ class Scratch3BoostBlocks {
         case BoostColorLabel.ANY:
             if (Object.keys(BoostColor).find(key => BoostColor[key])
                 .toLowerCase() !== this.getColor()) {
-                if (this.getColor() === this._peripheral.oldColor) {
```
Comment (evhan55) [inline src/extensions/scratch3_boost/index.js:1930]: Was this a bug before and should have been `this._peripheral._sensors.oldColor`? (Just trying to understand the change completely)

**Comment by evhan55** [comment]
The color name change should be okay, but I am not testing it since the color behavior is flaky at the moment in general for me, especially without LEGO bricks for testing.  I would say the color name change in this PR should be fine except for line #1929 which actually changes behavior a tiny bit.  If that is just a small bug fix, then it could land with this PR.  However, if it's important for me to test the change at line #1929, then I would maybe put all the color name change plus that bug fix in line #1929 in another PR later that addresses the color sensor changes if any will get made.  What do you think @knandersen ?

File: src/extensions/scratch3_boost/index.js (hunk 2 above)
Comment (knandersen) [reply src/extensions/scratch3_boost/index.js:1930]: yes, you're right, it should be `this._peripheral._sensors.previousColor`(also updated the naming). It was addressed in commit e3cdbff

**Comment by evhan55** [comment]
This feels mostly good to me now! 🎉 

- Can change motor direction for all motor on/seconds/rotations
- Can change motor power for all motor on/seconds/rotations
- Default was 50%
- Color bucket name change seems fine

Note:
- There is a tiny issue with rotations maybe not completing 100%, but I would rather find a repro for it and to resolve in a future PR, what do you think?

**Review by evhan55 — APPROVED** [review]
Looks mostly fine to me!  I think it might be good to merge at this step for testing at tomorrow's workshop.

**Comment by knandersen** [comment]
Thanks @evhan55!
I had a really good chat with @ericrosenbaum earlier today about cleaning up handling what state the motor is in, since the code right now is error-prone and very hard to read.
I've tried to put our thinking in writing:

Since the motors can either really be off, on forever, on for an amount of time, or on for an amount of rotation, the idea would be to create a BoostMotorStatus-enum, with these values:

```
BoostMotorStatus {
  OFF: 0
  ON_FOREVER: 1
  ON_FOR_TIME: 2
  ON_FOR_ROTATION: 3
}
```

`BoostMotor.isOn()` would be replaced by `BoostMotor.getStatus()` which would get `BoostMotor._status` which can have a value from the `BoostMotorStatus`-enum.
A new function called `BoostMotor.setStatus(boostMotorStatus)` would be created, which would be responsible for not only giving the motor a status, but depending on the value supplied to it will clear the relevant motor-states, i.e.:

```
setStatus(state) {
  switch(state){
    case BoostMotorState.OFF:
      // Clear timed- and rotation-based motor-states
    case BoostMotorState.ON_FOREVER:
      // Clear rotation-based and timed motor-states
    case BoostMotorState.ON_FOR_TIME:
      // Clear rotation-based motor-states
    case BoostMotorState.ON_FOR_ROTATION:
      // Clear timed motor-states
  }
}
```

Thoughts? cc @cwillisf

**Comment by evhan55** [comment]
@knandersen 

Clean up sounds good as long as it keeps the functionality.  It would be easier for me to see if it makes the code more legible once this proposed enum is actually implemented.

One extra thought: my opinion would be to land the functionality as is, to keep a snapshot of "working code" which was tricky to get right. Then, make a new PR with the clean up steps, hopefully feeling safer to do that refactor knowing that you can always revert to working code.  Just my two cents!
